pkg/azure/aks/machine_store.go

```diff
@@ -0,0 +1,273 @@
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/containerservice/armcontainerservice/v5"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
+
+	"github.com/Azure/go-autorest/autorest/to"
+)
+
+type VirtualMachineInfo struct {
+	Name            string
+	Region          string
+	OwningVMSS      string
+	OwningCluster   string
+	MachineTypeSku  string
+	OperatingSystem MachineOperatingSystem
+	Priority        MachinePriority
+
+	// TODO - decide if this should be stored here or not
+	// RetailPrice     float64
+}
+
+type MachineStore struct {
+	context context.Context
+	logger  *slog.Logger
+
+	azResourceGroupClient *armresources.ResourceGroupsClient
+	azVMSSClient          *armcompute.VirtualMachineScaleSetsClient
+	azVMSSVmClient        *armcompute.VirtualMachineScaleSetVMsClient
+	azAksClient           *armcontainerservice.ManagedClustersClient
+
+	RegionList     []string
+	regionListLock *sync.RWMutex
+
+	MachineMap     map[string]*VirtualMachineInfo
+	machineMapLock *sync.RWMutex
+}
+
+func NewMachineStore(parentCtx context.Context, parentLogger *slog.Logger, subscriptionId string, credentials *azidentity.DefaultAzureCredential) (*MachineStore, error) {
+	logger := parentLogger.With("subsystem", "machineStore")
+
+	rgClient, err := armresources.NewResourceGroupsClient(subscriptionId, credentials, nil)
+	if err != nil {
+		return nil, ErrClientCreationFailure
+	}
+
+	computeClientFactory, err := armcompute.NewClientFactory(subscriptionId, credentials, nil)
+	if err != nil {
+		return nil, ErrClientCreationFailure
+	}
+
+	containerClientFactory, err := armcontainerservice.NewClientFactory(subscriptionId, credentials, nil)
+	if err != nil {
+		return nil, ErrClientCreationFailure
+	}
+
+	ms := &MachineStore{
+		context: parentCtx,
+		logger:  logger,
+
+		azResourceGroupClient: rgClient,
+		azVMSSClient:          computeClientFactory.NewVirtualMachineScaleSetsClient(),
+		azVMSSVmClient:        computeClientFactory.NewVirtualMachineScaleSetVMsClient(),
+		azAksClient:           containerClientFactory.NewManagedClustersClient(),
+
+		RegionList:     []string{},
+		regionListLock: &sync.RWMutex{},
+
+		MachineMap:     make(map[string]*VirtualMachineInfo),
+		machineMapLock: &sync.RWMutex{},
+	}
+
+	go func() {
+		err := ms.PopulateMachineStore()
+		if err != nil {
+			ms.logger.LogAttrs(ms.context, slog.LevelError, "error populating initial machine store", slog.String("error", err.Error()))
+		}
+	}()
+
+	return ms, nil
+}
+
+func (m *MachineStore) getVmInfoByVmName(vmName string) *VirtualMachineInfo {
+	m.machineMapLock.RLock()
+	defer m.machineMapLock.RUnlock()
+
+	vmInfo := m.MachineMap[vmName]
+
+	return &VirtualMachineInfo{
+		Name:            vmInfo.Name,
+		Region:          vmInfo.Region,
+		OwningVMSS:      vmInfo.OwningVMSS,
+		MachineTypeSku:  vmInfo.MachineTypeSku,
+		OperatingSystem: vmInfo.OperatingSystem,
+		Priority:        vmInfo.Priority,
+	}
```

Comment:
Is there any reason not to store this as the object in the `m.MachineMap`? It's fine to create it each time, but seems like you could simply this greatly if it's stored this way.